tools/xc7frames2bit.cc

```diff
@@ -0,0 +1,398 @@
+	                  FLAGS_part_name.end());
+	bit_header.push_back(0x0);
+
+	// Build timestamp.
+	auto build_time = absl::Now();
+	auto build_date_string =
+	    absl::FormatTime("%E4Y/%m/%d", build_time, absl::UTCTimeZone());
+	auto build_time_string =
+	    absl::FormatTime("%H:%M:%S", build_time, absl::UTCTimeZone());
+
+	bit_header.push_back('c');
+	bit_header.push_back(
+	    static_cast<uint8_t>((build_date_string.size() + 1) >> 8));
+	bit_header.push_back(
+	    static_cast<uint8_t>(build_date_string.size() + 1));
+	bit_header.insert(bit_header.end(), build_date_string.begin(),
+	                  build_date_string.end());
+	bit_header.push_back(0x0);
+
+	bit_header.push_back('d');
+	bit_header.push_back(
+	    static_cast<uint8_t>((build_time_string.size() + 1) >> 8));
+	bit_header.push_back(
+	    static_cast<uint8_t>(build_time_string.size() + 1));
+	bit_header.insert(bit_header.end(), build_time_string.begin(),
+	                  build_time_string.end());
+	bit_header.push_back(0x0);
+
+	bit_header.insert(bit_header.end(), {'e', 0x0, 0x0, 0x0, 0x0});
+
+	return bit_header;
+}
+
+int writeBitstream(const ConfigurationPackets& packets,
+                   const std::string& output_file) {
+	std::ofstream out_file(output_file);
+	if (!out_file) {
+		std::cerr << "Unable to open file for writting: " << output_file
+		          << std::endl;
+		return 1;
+	}
+
+	BitHeader bit_header(createBitHeader());
+	out_file.write(reinterpret_cast<const char*>(bit_header.data()),
+	               bit_header.size());
+
+	auto end_of_header_pos = out_file.tellp();
+	auto header_data_length_pos =
+	    end_of_header_pos - static_cast<std::ofstream::off_type>(4);
+
+	xc7series::BitstreamWriter out_bitstream_writer(packets);
+	for (uint32_t word : out_bitstream_writer) {
+		out_file.put((word >> 24) & 0xFF);
+		out_file.put((word >> 16) & 0xFF);
+		out_file.put((word >> 8) & 0xFF);
+		out_file.put((word)&0xFF);
+	}
+
+	uint32_t length_of_data = out_file.tellp() - end_of_header_pos;
+
+	out_file.seekp(header_data_length_pos);
+	out_file.put((length_of_data >> 24) & 0xFF);
+	out_file.put((length_of_data >> 16) & 0xFF);
+	out_file.put((length_of_data >> 8) & 0xFF);
+	out_file.put((length_of_data)&0xFF);
+	return 0;
+}
+
+int main(int argc, char* argv[]) {
+	gflags::SetUsageMessage(argv[0]);
+	gflags::ParseCommandLineFlags(&argc, &argv, true);
+
+	auto part = xc7series::Part::FromFile(FLAGS_part_file);
+	if (!part) {
+		std::cerr << "Part file not found or invalid" << std::endl;
+		return 1;
+	}
+
+	FrameAddress2Frames frames;
+	if (FLAGS_frm_file.empty()) {
+		std::cerr << "Unable to open frames file: " << FLAGS_frm_file
+		          << std::endl;
+		return 1;
+	} else {
+		if (readFrames(FLAGS_frm_file, frames)) {
+			std::cerr << "Failed to read frames" << std::endl
+			          << "Exitting" << std::endl;
+			return 1;
+		}
+	}
+
+	// Put together a configuration package
+	ConfigurationPackets out_packets;
+
+	// Initialization sequence
```

Comment:
Can we link to where this sequence came from?

Reply:
The sequence came from https://www.kc8apf.net/2018/05/unpacking-xilinx-7-series-bitstreams-part-2/. It is exactly how it's done in xc7patch.

Reply:
If it is identical to how it is done in xc7patch, please make function, and in the function comment link to that page.